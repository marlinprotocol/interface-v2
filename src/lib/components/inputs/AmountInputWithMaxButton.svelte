<script lang="ts">
	import InputCard from '$lib/atoms/cards/InputCard.svelte';
	import Divider from '$lib/atoms/divider/Divider.svelte';
	import Text from '$lib/atoms/texts/Text.svelte';
	import { connected } from '$lib/data-stores/walletProviderStore';
	import type { InputCardVariant, ModalInputModel } from '$lib/types/componentTypes';
	import AmountInput from '$lib/components/inputs/AmountInput.svelte';
	import Tooltip from '$lib/atoms/tooltips/Tooltip.svelte';
	export let inputAmountString = '';
	export let maxAmountText: ModalInputModel['maxAmountText'] = 'Balance: 0.00';
	export let maxAmountTooltipText: ModalInputModel['maxAmountTooltipText'] = '';
	export let handleUpdatedAmount: any = undefined;
	export let inputCardVariant: InputCardVariant = 'primary';
	export let currency = '';
	export let showBalance: boolean = true;
	export let onlyInteger: boolean = false;
	export let disabled: boolean = false;
</script>

<InputCard variant="primary">
	<form>
		<div class="flex items-center justify-between gap-2">
			<AmountInput
				bind:value={inputAmountString}
				onChange={handleUpdatedAmount}
				styleClass="input input-ghost input-primary text-3xl  placeholder:text-grey-800/[.2]"
				disabled={!handleUpdatedAmount || !$connected || disabled}
				{onlyInteger}
			/>
			{#if currency !== ''}
				<span class="text-[22px] font-medium">{currency}</span>
			{/if}
		</div>
		{#if inputCardVariant !== 'none'}
			<Divider />
		{/if}
		{#if showBalance}
			<div class="mt-[-10px] flex items-center justify-end gap-2">
				<div class="flex h-8 items-center gap-1">
					<Text
						text={maxAmountText}
						variant="small"
						styleClass="text-[#657183]"
						fontWeight="font-normal"
					/>
					{#if maxAmountTooltipText}
						<Tooltip>
							<span slot="tooltipContent" class="text-sm leading-[18px]"
								>{maxAmountTooltipText}</span
							>
						</Tooltip>
					{/if}
				</div>
				{#if $$slots.inputMaxButton}
					<Divider direction="divider-vertical" />
					<slot name="inputMaxButton" />
				{/if}
			</div>
		{:else}
			<!-- this is added so that when balance is not being shown the height of the component remains consistent with the one that shows balance -->
			<div aria-disabled="true" class="h-[22px]"></div>
		{/if}
	</form>
</InputCard>
