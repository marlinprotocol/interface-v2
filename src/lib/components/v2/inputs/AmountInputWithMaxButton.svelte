<script lang="ts">
	import InputCard from '$lib/atoms/v2/cards/InputCard.svelte';
	import Divider from '$lib/atoms/v2/divider/Divider.svelte';
	import Text from '$lib/atoms/v2/texts/Text.svelte';
	import { connected } from '$lib/data-stores/walletProviderStore';
	import type { InputCardVariant, ModalInputModel } from '$lib/types/componentTypes';
	import AmountInput from '$lib/components/v2/inputs/AmountInput.svelte';
	import Tooltip from '$lib/atoms/v2/tooltips/Tooltip.svelte';
	export let inputAmountString = '';
	export let maxAmountText: ModalInputModel['maxAmountText'] = 'Balance: 0.00';
	export let maxAmountTooltipText: ModalInputModel['maxAmountTooltipText'] = '';
	export let handleUpdatedAmount: any = undefined;
	export let inputCardVariant: InputCardVariant = 'primary';
	export let currency = '';
	export let showBalance: boolean = true;
</script>

<InputCard variant="primary">
	<form>
		<div class="flex items-center justify-between gap-2">
			<AmountInput
				bind:value={inputAmountString}
				onChange={handleUpdatedAmount}
				styleClass="input input-ghost input-primary text-3xl placeholder:text-[#030115]/[.2]"
				disabled={!handleUpdatedAmount || !$connected}
			/>
			{#if currency !== ''}
				<span class="text-[22px] font-medium">{currency}</span>
			{/if}
		</div>
		{#if inputCardVariant !== 'none'}
			<Divider />
		{/if}
		{#if showBalance}
			<div class="mt-[-10px] flex items-center justify-end gap-2">
				<div class="flex items-center gap-1">
					<Text
						text={maxAmountText}
						variant="small"
						styleClass="text-[#657183]"
						fontWeight="font-normal"
					/>
					{#if maxAmountTooltipText}
						<Tooltip>
							<span slot="tooltipContent" class="text-sm leading-[18px]"
								>{maxAmountTooltipText}</span
							>
						</Tooltip>
					{/if}
				</div>
				{#if $$slots.inputMaxButton}
					<Divider direction="divider-vertical" />
					<slot name="inputMaxButton" />
				{/if}
			</div>
		{/if}
	</form>
</InputCard>
