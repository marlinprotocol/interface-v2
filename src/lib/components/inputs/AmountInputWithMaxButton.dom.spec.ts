import { render, screen, fireEvent } from '@testing-library/svelte';
import { describe, it, expect, vi } from 'vitest';
import AmountInputWithMaxButton from './AmountInputWithMaxButton.svelte'; // Update with the actual path
import { tick } from 'svelte';
import html from '@playpilot/svelte-htm';
const { connected } = await vi.hoisted(
	() => import('$lib/data-stores/mock-data-stores/mockStores')
);

beforeAll(() => {
	vi.mock('$lib/data-stores/walletProviderStore', async () => {
		return {
			connected
		};
	});
});

describe('AmountInputWithMaxButton', () => {
	const tooltipText = 'Info Tooltip';
	const inputAmountString = '123';
	const maxAmountText = 'Balance: 1000';
	const maxAmountTooltipText = 'Your total balance';
	const handleUpdatedAmount = vi.fn();

	it("render's the component properly", () => {
		const { container } = render(AmountInputWithMaxButton, {
			inputAmountString,
			maxAmountText,
			maxAmountTooltipText,
			handleUpdatedAmount,
			inputCardVariant: 'primary'
		});
		expect(container).toMatchSnapshot();
	});

	it('renders correctly based on props and slots', async () => {
		const { getByDisplayValue, container } = render(AmountInputWithMaxButton, {
			inputAmountString,
			maxAmountText,
			maxAmountTooltipText,
			handleUpdatedAmount,
			inputCardVariant: 'primary'
		});

		// Assert that title and tooltip are rendered
		expect(container.innerHTML).contain(tooltipText);

		// Assert input is bound to inputAmountString
		const input = getByDisplayValue(inputAmountString);
		expect(input).toBeTruthy();

		// Simulate user input
		fireEvent.input(input, { target: { value: '456' } });
		expect(handleUpdatedAmount).toHaveBeenCalled();
	});

	it('disables the input when handleUpdatedAmount is undefined or connected is false', async () => {
		connected.mockSetSubscribeValue(false);
		const { getByDisplayValue } = render(AmountInputWithMaxButton, {
			inputAmountString: '0',
			handleUpdatedAmount: undefined // Not providing handleUpdatedAmount makes the input disabled
		});

		const input = getByDisplayValue('0') as HTMLInputElement;
		expect(input.disabled).toBeTruthy();
	});

	it('enables the input when handleUpdatedAmount is undefined or connected is true', async () => {
		// Update `connected` store
		connected.mockSetSubscribeValue(true);
		await tick();
		const { getByPlaceholderText } = render(AmountInputWithMaxButton, {
			inputAmountString: '0',
			handleUpdatedAmount: vi.fn() // Providing this should enable the input
		});

		expect((getByPlaceholderText('0.00') as HTMLInputElement).disabled).toBeFalsy();
	});

	it('check that Divider is not rendered without inputMaxButton slot', () => {
		const divider = screen.queryByTestId('divider');
		expect(divider).not.toBeTruthy();
	});

	it('check that Divider is rendered withinputMaxButton slot', async () => {
		const { getAllByTestId, getByText } = render(html`
        <${AmountInputWithMaxButton} title='${'text'}' inputAmountString=${'0'} handleUpdatedAmount=${vi.fn()}>
        <div slot='inputMaxButton'>inputMaxButton</div>
        <div slot='input-end-button'>input-end-button</div>
        </${AmountInputWithMaxButton}>`);
		const dividers = getAllByTestId('divider');
		expect(dividers.length).toBeTruthy();
		expect(dividers.length).toBe(2);
		expect(await getByText('inputMaxButton')).toBeTruthy();
		expect(await getByText('input-end-button')).toBeTruthy();
	});
});
