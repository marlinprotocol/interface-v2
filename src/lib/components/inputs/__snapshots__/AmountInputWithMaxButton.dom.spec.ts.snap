// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AmountInputWithMaxButton > render's the component properly 1`] = `
<div>
  <div
    class="w-full bg-white px-6 py-[30px] rounded-3xl border border-[#D9DADE]"
    data-testid="input-card"
  >
    <form>
      <div
        class="flex items-center justify-between gap-2"
      >
        <input
          autocomplete="off"
          class="hideInputNumberAppearance input input-ghost input-primary text-3xl placeholder:text-[#030115]/[.2]"
          disabled=""
          id="pond-input-amount-"
          placeholder="0.00"
        />
        <!--&lt;AmountInput&gt;-->
         
      </div>
       
      <div
        class="h-[1px] w-full bg-black/[0.1] my-0 shrink-0"
        data-testid="divider"
      />
      <!--&lt;Divider&gt;-->
       
      <div
        class="mt-[-10px] flex items-center justify-end gap-2"
      >
        <div
          class="flex items-center gap-1"
        >
          <div
            class="text-[#657183] text-sm font-normal"
            data-testid="text"
          >
            Balance: 1000
          </div>
          <!--&lt;Text&gt;-->
           
          <div
            class="group relative inline-block cursor-pointer text-left"
            data-testid="tooltip"
          >
            <img
              alt="Info"
              src="/images/alertV2Icon.svg"
              width="16px"
            />
             
             
            <div
              class="absolute z-100 box-border hidden w-max max-w-[320px] rounded-xl bg-white p-3 shadow-[0px_12px_16px_-4px_#10182814] group-hover:block -top-5 left-1/2 -translate-x-1/2 -translate-y-full"
            >
              <span
                class="text-sm leading-[18px]"
                slot="tooltipContent"
              >
                Your total balance
              </span>
               
              <i
                class="absolute overflow-hidden after:absolute after:h-[15px] after:w-[15px] after:rotate-45 after:bg-white after:shadow-[0px_12px_16px_-4px_#10182814] after:left-1/2 after:-translate-x-1/2 after:-translate-y-1/2 top-full left-1/2 -ml-[15px] w-[30px] h-[15px]"
              />
               
              <div
                class="after:left-0 after:top-1/2 after:-translate-y-1/2 after:translate-x-1/2"
              />
            </div>
          </div>
          <!--&lt;Tooltip&gt;-->
        </div>
         
      </div>
    </form>
  </div>
  <!--&lt;InputCard&gt;-->
  <!--&lt;AmountInputWithMaxButton&gt;-->
</div>
`;
